### PR TITLE
porting appcheck with C#

### PR DESCRIPTION
**Pull request by taeul**
Ported the Node.js appcheck SDK with C#

### CODE REVIEW

**Comment by taeul** [comment]
Implement migrating and testing

File: FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckApiClient.cs [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+    /// </summary>
+    /// <returns>A task that completes with the creation of a new App Check token.</returns>
+    /// <exception cref="ArgumentNullException">Thrown if an error occurs while creating the custom token.</exception>
+    /// <value>The Firebase app instance.</value>
+    internal class AppCheckApiClient
+    {
+        private const string ApiUrlFormat = "https://firebaseappcheck.googleapis.com/v1/projects/{projectId}/apps/{appId}:exchangeCustomToken";
+        private const string OneTimeUseTokenVerificationUrlFormat = "https://firebaseappcheck.googleapis.com/v1beta/projects/{projectId}:verifyAppCheckToken";
+        private readonly FirebaseApp app;
+        private string projectId;
```
Comment (jopmiddelkamp) [inline FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckApiClient.cs]:
```suggestion
        private string _projectId;
```

Comment (jopmiddelkamp) [inline FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckApiClient.cs]: Why do you set this project id like this while the JS SDK does some kind of async check?

```js
  private getProjectId(): Promise<string> {
    if (this.projectId) {
      return Promise.resolve(this.projectId);
    }
    return utils.findProjectId(this.app)
      .then((projectId) => {
        if (!validator.isNonEmptyString(projectId)) {
          throw new FirebaseAppCheckError(
            'unknown-error',
            'Failed to determine project ID. Initialize the '
            + 'SDK with service account credentials or set project ID as an app option. '
            + 'Alternatively, set the GOOGLE_CLOUD_PROJECT environment variable.');
        }
        this.projectId = projectId;
        return projectId;
      });
  }
```

File: FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckApiClient.cs
Comment (jopmiddelkamp) [inline FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckApiClient.cs:1]: I'm missing the `FirebaseError` here but also in some other classes.

File: FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckApiClient.cs [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+        /// <summary>
+        /// Get url from FirebaseApp Id .
+        /// </summary>
+        /// <param name="appId">The FirebaseApp Id.</param>
+        /// <returns>A formatted verify token url.</returns>
+        private string GetUrl(string appId)
+        {
+            if (string.IsNullOrEmpty(this.projectId))
+            {
+                this.projectId = this.app.GetProjectId();
```
Comment (jopmiddelkamp) [inline FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckApiClient.cs]: This feels weird why do we eve have the `projectId` defined as a field?

File: FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckApiClient.cs [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+            return int.Parse(modifiedString) * 1000;
+        }
+
+        /// <summary>
+        /// Formats a string of form 'project/{projectId}/{api}' and replaces with corresponding arguments {projectId: '1234', api: 'resource'}.
+        /// </summary>
+        /// <param name="str">The original string where the param need to be replaced.</param>
+        /// <param name="urlParams">The optional parameters to replace in thestring.</param>
+        /// <returns> The resulting formatted string. </returns>
+        private string FormatString(string str, Dictionary<string, string> urlParams)
```
Comment (jopmiddelkamp) [inline FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckApiClient.cs]: This is such a generic method that has nothing to do with the app check api client. Please move this to a seperate util class.

File: FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckToken.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+namespace FirebaseAdmin.Check
+{
+    /// <summary>
+    /// Interface representing an App Check token.
+    /// </summary>
+    /// <remarks>
+    /// Initializes a new instance of the <see cref="AppCheckToken"/> class.
+    /// </remarks>
+    /// <param name="tokenValue">Generator from custom token.</param>
```
Comment (jopmiddelkamp) [inline FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckToken.cs:6]: Is this description correct? It is not matching the description of the property.

File: FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckTokenVerify.cs [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+{
+    internal class AppCheckTokenVerify
+    {
+        /*private const string IdTokenCertUrl = "https://www.googleapis.com/robot/v1/metadata/x509/"
+            + "securetoken@system.gserviceaccount.com";
+
+        private const string SessionCookieCertUrl = "https://www.googleapis.com/identitytoolkit/v3/"
+            + "relyingparty/publicKeys";
+*/
+        private const string AppCheckIuuser = "https://firebaseappcheck.googleapis.com/";
```
Comment (jopmiddelkamp) [inline FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckTokenVerify.cs]:
```suggestion
        private const string AppCheckIssuer = "https://firebaseappcheck.googleapis.com/";
```

File: FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckTokenVerify.cs [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+using FirebaseAdmin.Auth;
+using FirebaseAdmin.Auth.Jwt;
+using Google.Apis.Auth;
+using Google.Apis.Util;
+
+namespace FirebaseAdmin.Check
+{
+    internal class AppCheckTokenVerify
+    {
+        /*private const string IdTokenCertUrl = "https://www.googleapis.com/robot/v1/metadata/x509/"
```
Comment (jopmiddelkamp) [inline FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckTokenVerify.cs]: Why is this here? If we don't use it why don't we remove it?

File: FirebaseAdmin/FirebaseAdmin/Auth/FirebaseToken.cs [hunk 7]
```diff
@@ -80,9 +87,20 @@ internal FirebaseToken(Args args)
         /// </summary>
         public IReadOnlyDictionary<string, object> Claims { get; }
 
+        /// <summary>
+        /// Defined operator string.
+        /// </summary>
+        /// <param name="v">FirebaseToken.</param>
+        public static implicit operator string(FirebaseToken v)
+        {
+            throw new NotImplementedException();
```
Comment (jopmiddelkamp) [inline FirebaseAdmin/FirebaseAdmin/Auth/FirebaseToken.cs]: Please implement this.

Comment (jopmiddelkamp) [inline FirebaseAdmin/FirebaseAdmin/Auth/FirebaseToken.cs]: Please implement this.

File: FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckTokenVerify.cs [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using FirebaseAdmin.Auth;
+using FirebaseAdmin.Auth.Jwt;
+using Google.Apis.Auth;
+using Google.Apis.Util;
+
+namespace FirebaseAdmin.Check
+{
+    internal class AppCheckTokenVerify
```
Comment (jopmiddelkamp) [inline FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckTokenVerify.cs]:
```suggestion
    internal class AppCheckTokenVerifier
```

File: FirebaseAdmin/FirebaseAdmin/Auth/Jwt/FirebaseTokenFactory.cs [hunk 9]
```diff
@@ -173,14 +176,75 @@ internal static FirebaseTokenFactory Create(FirebaseApp app, string tenantId = n
                 header, payload, this.Signer, cancellationToken).ConfigureAwait(false);
         }
 
+        internal async Task<string> CreateCustomTokenAppIdAsync(
```
Comment (jopmiddelkamp) [inline FirebaseAdmin/FirebaseAdmin/Auth/Jwt/FirebaseTokenFactory.cs]: I feel like this should be isolated into the `FirebaseAdmin.AppCheck` namespace. Why did you put app check specific code into the `FirebaseAdmin.Auth` namespace? I don't see anything `AppCheck` related in the `auth` directory in the JS SDK either.

**Review by jopmiddelkamp — CHANGES_REQUESTED** [review]

File: FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckApiClient.cs (hunk 1 above)
Comment (rising-star92) [reply FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckApiClient.cs:27]: "_projectId" is not allowed by naming rule on this project. So "projectId" is correct.

Comment (taeul) [reply FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckApiClient.cs]: "_projectId" is not allowed by naming rule on this project. So I think that "projectId" is correct.

Comment (taeul) [reply FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckApiClient.cs]: In JS SDK, Async check is used for Flexibility, Environment Variables and Metadata Service, Caching the Result
When running on Google Cloud Platform (GCP), if the project ID is not explicitly set, it is needed

In contrast, in a C# environment (typically server-side), the common practice is to initialize the SDK with explicit configuration, often including the project ID. The configuration is usually loaded synchronously during application startup. The project ID might come from a configuration file, environment variable, or explicitly passed in when initializing the SDK. As a result, an asynchronous check for the project ID is usually not necessary.

File: FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckApiClient.cs
Comment (taeul) [reply FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckApiClient.cs:1]: Thank you for bringing this to my attention. I understand the importance of consistent and robust error handling in our application, especially when interacting with external services like Firebase. To address this, I have created a custom FirebaseAppCheckException class that encapsulates Firebase-specific error information

File: FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckApiClient.cs (hunk 2 above)
Comment (taeul) [reply FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckApiClient.cs]: I have updated the AppCheckApiClient.cs file, and the issue has been resolved

File: FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckApiClient.cs (hunk 3 above)
Comment (taeul) [reply FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckApiClient.cs]: I have updated the AppCheckApiClient.cs file and move this function to util, and the issue has been resolved

File: FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckTokenVerify.cs (hunk 5 above)
Comment (taeul) [reply FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckTokenVerify.cs]: I have updated the AppCheckTokenVerifier.cs file, and the issue has been resolved.

File: FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckTokenVerify.cs (hunk 6 above)
Comment (taeul) [reply FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckTokenVerify.cs]: I have updated the AppCheckTokenVerifier.cs file, and the issue has been resolved.

File: FirebaseAdmin/FirebaseAdmin/Auth/FirebaseToken.cs (hunk 7 above)
Comment (taeul) [reply FirebaseAdmin/FirebaseAdmin/Auth/FirebaseToken.cs]: I have updated the auth/FirebaseToken.cs file, and the issue has been resolved.

Comment (taeul) [reply FirebaseAdmin/FirebaseAdmin/Auth/FirebaseToken.cs]: I have updated the auth/FirebaseToken.cs file, and the issue has been resolved.

File: FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckTokenVerify.cs (hunk 8 above)
Comment (taeul) [reply FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckTokenVerify.cs]: I have updated file and class name, and the issue has been resolved.

File: FirebaseAdmin/FirebaseAdmin/Auth/Jwt/FirebaseTokenFactory.cs (hunk 9 above)
Comment (taeul) [reply FirebaseAdmin/FirebaseAdmin/Auth/Jwt/FirebaseTokenFactory.cs]: I have updated code and the issue has been resolved.

File: FirebaseAdmin/FirebaseAdmin/Util/HttpUtils.cs [hunk 10]
```diff
@@ -29,6 +30,17 @@ internal sealed class HttpUtils
 
         private HttpUtils() { }
 
+        public static string FormatString(string str, Dictionary<string, string> urlParams)
```
Comment (jopmiddelkamp) [inline FirebaseAdmin/FirebaseAdmin/Util/HttpUtils.cs]: A format string method on a http utillity doesn't really explain what its doing right? I feel a name like `StringUtils.ReplacePlaceholders(...)` would make more sense.

File: FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckApiClient.cs [hunk 11]
```diff
@@ -1,198 +1,288 @@
...
-        /// Initializes a new instance of the <see cref="AppCheckApiClient"/> class.
-        /// </summary>
-        /// <param name="value"> Initailize FirebaseApp. </param>
-        public AppCheckApiClient(FirebaseApp value)
+        private readonly ErrorHandlingHttpClient<FirebaseAppCheckException> httpClient;
+        private readonly string projectId;
+
+        internal AppCheckApiClient(Args args)
+        {
+            string noProjectId = "Project ID is required to access app check service. Use a service account "
```
Comment (jopmiddelkamp) [inline FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckApiClient.cs]: Why putting it here as a string and not just in the if statement where it goes wrong?

**Review by jopmiddelkamp — CHANGES_REQUESTED** [review]

File: FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckApiClient.cs (hunk 2 above)
Comment (jopmiddelkamp) [reply FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckApiClient.cs]: What did you change exactly?

File: FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckToken.cs (hunk 4 above)
Comment (taeul) [reply FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckToken.cs:6]: I have updated the AppCheckToken.cs file, and the issue has been resolved

File: FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckApiClient.cs (hunk 11 above)
Comment (taeul) [reply FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckApiClient.cs]: I think that it is good suggestion, I move string in if statement.

File: FirebaseAdmin/FirebaseAdmin/Util/HttpUtils.cs (hunk 10 above)
Comment (taeul) [reply FirebaseAdmin/FirebaseAdmin/Util/HttpUtils.cs]: I think that it is good suggestion, I make StringUtils class and define ReplacePlaceholders() function.

File: FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckApiClient.cs (hunk 2 above)
Comment (taeul) [reply FirebaseAdmin/FirebaseAdmin/AppCheck/AppCheckApiClient.cs]: In AppCheckApiClient.cs file, projectId field is needed to generate exchangeTokenurl
So AppCheckApiClient get projectId from args.

**Review by jopmiddelkamp — APPROVED** [review]
